File: old_src_python/cadencement_opti.py
'''
Algorithme de cadencement de clé avec options de debugage.
'''
def keygen_opti(maitre, nb_tour, box_s):
    # Liste des sous clés.
    list_tour = []
    # Les masque a utiliser pendant les opérations.
    # Masque pour récupérer la clé du tour.
    masque_ct = 0xffffff
    # Masque à appliquer lors de la rotation.
    masque_rot = 0xffffffffffffffffffff
    # Masque à appliquer lors de la récupération des 4 bits de poids forts.
    masque_bitpf = 0xfffffffffffffffffff
    # Masque pour récupérer la partie à xorer
    masque_xor = 0x1f
    # Masque pour placer la partie xorée
    masque_xor2 = 0xfffffffffffffff07fff
    # On initialise le registre en plaçant la clé maitre dans les 24 bits de poids forts.
    reg = 0 | (maitre << 56 )
    for ind in range(1, nb_tour + 1):

        # On récupère la clé de ce tour et on le met dans la liste.
        cletour = reg >> 16 & masque_ct
        list_tour.append(cletour)
        a = 2
        a << 1
        # ETAPE 1: On pivote de 61 bits vers la gauches
        reg, reg2 = reg << 61 , reg >> (80 - 61)
        reg = reg & masque_rot | reg2

        # ETAPE 2: On fais passer les 4 bits de poids fort dans la boite S.
        bitpf = reg >> 76
        nvbitpf = box_s[bitpf] << 76
        reg = nvbitpf | (reg & masque_bitpf)

        # ETAPE 3: On xore le nombre formé par les 5 bit entre 19 et 15 avec le numero du round
        nb = (reg >> 15) & masque_xor
        nvlnb = (nb ^ ind) 
        reg = (nvlnb << 15) | (reg & masque_xor2)
    return list_tour

File: src/cadencement_opti.c
#include <stdint.h>
#define NB_TOUR 11

/*
Fonction qui applique l'algorithme de cadencement présenté dans l'énoncé.
*/
void keygen_opti(uint64_t maitre, int *list_tour, int* box_s){
    //Le maximum de bit d'un entier est de 64.
    //Il faut donc diviser en deux partie: 
    //part1 pour les bits de poids fort de 79 à 40.    
    uint64_t part1 = maitre << 16;
    //part2 pour les bits de poids faibles de 39 à 0.
    uint64_t part2 = 0;

    // Masque pour récupèrer les 19 bits de poids faibles. 
    uint64_t masque_pfa = 0x7ffff;
    // Masque pour supprimer les 4 bits de poids forts.
    uint64_t masque_4pfo = 0xfffffffff;

    uint64_t part1temp;
    uint64_t part2temp;
    
    for (int i = 1; i < NB_TOUR + 1; i++){
        // On récupère les clés de chaques tours.
        list_tour[i - 1] = part2 >> 16;

        // ETAPE 1: On pivote le registre de 61 bits vers la gauches 
        // Cela reviens à faire une rotation de 19 bits vers la droite.
        part1temp = (part1 & masque_pfa) << 21;
        part2temp = (part2 & masque_pfa) << 21;
        part1 = (part1 >> 19) | part2temp;
        part2 = (part2 >> 19) | part1temp;
        

        // ETAPE 2: On isole les 4 bits de poids fort pour les passer dans la box S.
        uint64_t bitpfo = part1 >> 36;
        // On supprimes les 4 bits de poids forts.
        part1 = part1 & masque_4pfo;
        uint64_t nvval = box_s[bitpfo];
        bitpfo = nvval << 36;
        part1 = part1 | bitpfo;

        // ETAPE 3: On xor le numéro de tour actuel au 5 bit entre 15 et 19. 
        uint64_t val = i << 15;
        part2 = part2 ^ val;
    }

}